### PR TITLE
Convert otel-arrow-adapter to use OTel Collector `pdata`

### PR DESCRIPTION
**Pull request by jmacd**
Replace direct use of OTLP protocol buffers with OTel Collector pdata objects.

This work was summarized in https://github.com/open-telemetry/opentelemetry-collector/issues/6166. The `pdata` API forces an OTLP+Arrow conversion to copy data that would otherwise be shared between rows. These be found in the current PR everywhere there is a arrow-to-OTLP conversion: once for the Resource object, once for the Scope object, and once for the Attributes object. These are places where an "unsafe" package could be introduced to allow receivers to intentionally share pointers, but for now this is regarded as a low priority.

## Draft Review notes

~This is a work-in-progress to convert @lquerel's OTLP-to-Arrow adapter (i.e., this repository) described in https://github.com/open-telemetry/oteps/pull/171, meant for "Phase 1" of the project which would modify the collector's core OTLP receiver and exporter with an option to use a column-oriented encoding based on [Apache Arrow](https://arrow.apache.org/).~

~This is a DRAFT meant for show-and-tell: it is currently incomplete and meant to provide feedback on `pdata`.~

### CODE REVIEW

File: pkg/otel/common/arrow_to_otlp.go [hunk 1]
```diff
@@ -19,261 +19,242 @@ package common
...
 		}
 		values, err := ArrayValueFrom(arrList, row)
 		if err != nil {
-			return nil, err
+			return err
 		}
-		return &commonpb.AnyValue{Value: &commonpb.AnyValue_ArrayValue{ArrayValue: &commonpb.ArrayValue{
-			Values: values,
-		}}}, nil
+		values.CopyTo(dest.SetEmptySliceVal())
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply pkg/otel/common/arrow_to_otlp.go:216]: Even better pass the `dest.SetEmptySliceVal()` to the `Array...` func and populate directly there.

**Comment by jmacd** [comment]
@bogdandrutu Thanks. I'm aware that I can usually rearrange the code not to require MoveTo() or Set. Since I'm refactoring a large code base, I've only partly completed that change, but I understand what you mean.

However, the `MoveTo()` only works if I'm not going to re-use the object, which is not true in some of these cases.

File: pkg/otel/common/arrow_to_otlp.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (jmacd) [reply pkg/otel/common/arrow_to_otlp.go]: Fixed.

File: pkg/benchmark/dataset/real_logs_dataset.go [hunk 2]
```diff
@@ -21,21 +21,19 @@ import (
...
 	logs []logUnit
 }
 
 type logUnit struct {
-	logRecord *logspb.LogRecord
-	resource  *logspb.ResourceLogs
-	scope     *logspb.ScopeLogs
+	logRecord plog.LogRecord
+	resource  plog.ResourceLogs
+	scope     plog.ScopeLogs
```
Comment (jmacd) [inline pkg/benchmark/dataset/real_logs_dataset.go:36]: Note that `plog` objects contain pointers internally, which is why they still work as map keys the way the `real_xxx_dataset.go` files use them.

File: pkg/datagen/default.go [hunk 3]
```diff
@@ -17,194 +17,128 @@ package datagen
...
 var HOSTNAMES = []string{"host1.mydomain.com", "host2.org", "host3.thedomain.edu", "host4.gov", "host5.retailer.com"}
 var UPS = []bool{true, false}
 var STATUS = []int64{200, 300, 400, 404, 500, 503}
 var VERSIONS = []string{"1.0.0", "1.0.2", "2.0", "1.9.9"}
 var STATES = []string{"running", "ready", "maintenance", "degraded", "unavailable", "unknown"}
-var TRACE_IDS = []string{"trace1", "trace2", "trace3", "trace4", "trace5"}
+var GROUP_IDS = []string{"group1", "group2", "group3", "group4", "group5"}
+
+type Attrs = pcommon.Map
+type AttrFunc func(Attrs)
```
Comment (jaronoff97) [inline pkg/datagen/default.go:31]: should this type be used in shuffleAttrs? If not, I think it's unused right now.

File: pkg/otel/metrics/otlp_to_arrow.go [hunk 4]
```diff
@@ -92,24 +100,26 @@ func OtlpMetricsToArrowRecords(rr *air.RecordRepository, request *collogspb.Expo
 	return result, nil
 }
 
-func addGaugeOrSum(rr *air.RecordRepository, resMetrics *metricspb.ResourceMetrics, scopeMetrics *metricspb.ScopeMetrics, metric *metricspb.Metric, dataPoints []*metricspb.NumberDataPoint, metric_type string, config *MultivariateMetricsConfig, cfg *config.Config) error {
-	if mvKey, ok := config.Metrics[metric.Name]; ok {
-		return multivariateMetric(rr, resMetrics, scopeMetrics, metric, dataPoints, metric_type, mvKey, cfg)
+func addGaugeOrSum(rr *air.RecordRepository, res pcommon.Resource, scope pcommon.InstrumentationScope, metric pmetric.Metric, dataPoints pmetric.NumberDataPointSlice, metric_type string, config *MultivariateMetricsConfig, cfg *config.Config) error {
+	// TODO: Missing Temporality and IsMonotonic for Sums here.
```
Comment (jaronoff97) [inline pkg/otel/metrics/otlp_to_arrow.go:104]: What happens to temporality and monotonicity currently? Do we just assume Delta and true?

File: pkg/otel/metrics/otlp_to_arrow.go [hunk 5]
```diff
@@ -168,15 +170,15 @@ func multivariateMetric(rr *air.RecordRepository, resMetrics *metricspb.Resource
...
 			continue
 		}
-		metricField := rfield.NewStructField(metric.Name, rfield.Struct{
+		metricField := rfield.NewStructField(metric.Name(), rfield.Struct{
 			Fields: record.metrics,
 		})
 		metricField.AddMetadata(constants.METADATA_METRIC_TYPE, metric_type)
-		if len(metric.Description) > 0 {
-			metricField.AddMetadata(constants.METADATA_METRIC_DESCRIPTION, metric.Description)
+		if metric.Description() != "" {
```
Comment (jaronoff97) [inline pkg/otel/metrics/otlp_to_arrow.go:177]: nit: should we keep the len check for the string?

**Review by jaronoff97 — APPROVED** [review]
A few non-blocking comments, overall this makes sense.

File: pkg/otel/metrics/otlp_to_arrow.go (hunk 4 above)
Comment (lquerel) [reply pkg/otel/metrics/otlp_to_arrow.go:104]: Indeed we need to clarify that.

File: pkg/otel/metrics/otlp_to_arrow.go (hunk 5 above)
Comment (lquerel) [reply pkg/otel/metrics/otlp_to_arrow.go:177]: No opinion on that. IMO, we take the approach that is most effective.

**Review by lquerel — APPROVED** [review]
LGTM. 

Big thanks for this integration with `pdata` and your different modifications to make this code a little more idiomatic with Golang.

**Pull request merged**